File: src/main/java/cs/android/viewbase/menu/CSOnMenu.java
package cs.android.viewbase.menu;

import androidx.appcompat.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;

import cs.java.lang.CSValue;

import static cs.java.lang.CSLang.NO;
import static cs.java.lang.CSLang.YES;
import static cs.java.lang.CSLang.set;

public class CSOnMenu {

    private final Menu _menu;
    private final CSValue<Boolean> _showMenu;
    private final AppCompatActivity _activity;

    public CSOnMenu(AppCompatActivity activity, Menu menu) {
        _activity = activity;
        _menu = menu;
        _showMenu = new CSValue<>(false);
    }

    public MenuItem find(int id) {
        return _menu.findItem(id);
    }

    public CSOnMenu hide(int id) {
        find(id).setVisible(NO);
        return this;
    }

    public void show(CSMenuItem item) {
        if (set(item.id())) {
            MenuItem menuItem = find(item.id());
            menuItem.setVisible(YES);
            if (menuItem.isCheckable()) menuItem.setChecked(item.isChecked());
        } else {
            MenuItem menuItem = _menu.add(item.title());
            menuItem.setShowAsAction(item.showAsAction());
            if (menuItem.isCheckable()) menuItem.setChecked(item.isChecked());
            if (set(item.icon())) menuItem.setIcon(item.icon());
        }
    }

    public boolean showMenu() {
        return _showMenu.getValue();
    }

    public void showMenu(boolean showMenu) {
        _showMenu.setValue(showMenu);
    }

    public void inflate(int menuId) {
        _activity.getMenuInflater().inflate(menuId, _menu);
        showMenu(YES);
    }

    public Menu getMenu() {
        return _menu;
    }
}

File: library/src/main/java/renetik/android/viewbase/menu/CSOnMenu.kt
package renetik.android.viewbase.menu

import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import renetik.android.java.collections.CSList
import renetik.android.java.lang.CSValue
import renetik.android.lang.CSLang.YES
import renetik.android.lang.CSLang.set

const val GeneratedMenuItems = 7687678

class CSOnMenu(private val activity: AppCompatActivity, val menu: Menu) {

    val showMenu = CSValue(true)

    fun find(id: Int): MenuItem? = menu.findItem(id)

    fun show(item: CSMenuItem) {
        if (item.isGenerated) createProgrammaticallyItem(item)
        else find(item.id)?.apply {
            isVisible = YES
            item.title?.let { title = it }
            if (isCheckable) item.isChecked?.let { isChecked = it }
                    ?: let { item.isChecked = isChecked }
            actionView?.let { item.actionView }
        }
    }

    private fun createProgrammaticallyItem(item: CSMenuItem) {
        menu.add(GeneratedMenuItems, item.id, Menu.NONE, item.title).apply {
            setShowAsAction(item.showAsAction)
            if (set(item.iconResource)) setIcon(item.iconResource)
            item.isChecked?.let {
                isCheckable = true
                isChecked = it
            }
            item.actionView?.let { actionView = it }
        }
    }

    fun inflate(menuId: Int) = activity.menuInflater.inflate(menuId, menu)

    fun <T : CSMenuItem> onPrepareItems(menuItems: CSList<T>) {
        for (item in menuItems) if (item.isVisible) show(item)
    }

}